eng/scripts/Language-Settings.ps1

```diff
@@ -10,54 +10,187 @@ $PackageRepositoryUri = "https://repo1.maven.org/maven2"
 
-function Get-java-PackageInfoFromRepo ($pkgPath, $serviceDirectory)
-{
-  $projectPath = Join-Path $pkgPath "pom.xml"
-  if (Test-Path $projectPath)
-  {
-    $projectData = New-Object -TypeName XML
-    $projectData.load($projectPath)
+# When getting all of the package properties, if Get-AllPackageInfoFromRepo exists
+# then it's called instead of Get-PkgPropsForEntireService.
+function Get-AllPackageInfoFromRepo([string]$serviceDirectory = $null) {
+  $SdkType = $Env:SdkType
+  if ($SdkType) {
+    Write-Host "SdkType env var was set to '$SdkType'"
+  } else {
+    $SdkType = "client"
+    Write-Host "SdkType env var was not set, default to 'client'"
+  }
+  Write-Host "Processing SdkType=$SdkType"
+
+  $allPackageProps = @()
+  $sdkRoot = Join-Path $RepoRoot "sdk"
+  $ymlFiles = @()
+
+  if ($serviceDirectory) {
+    $searchPath = Join-Path $sdkRoot $serviceDirectory
+    Write-Host "searchPath=$searchPath"
+    [array]$ymlFiles = Get-ChildItem -Path $searchPath "ci*.yml" | Where-Object { $_.PSIsContainer -eq $false}
+  } else {
+    # The reason for the exclude folders are POM only releases (nothing is built) or
+    # the service directory sits outside of the engineering system
+    # 1. boms - BOMs are POM only releases. Also, their versions aren't version controlled.
+    # 2. parents - parents are POM only releases which are version controlled
+    # 3. resourcemanagerhybrid - intermediate version of resourcemanager that was
+    #    a one time release which sits outside of the engineering system
+    $excludeFolders = "boms", "resourcemanagerhybrid", "parents"
+    [array]$ymlFiles = Get-ChildItem -Path $sdkRoot -Include "ci*.yml" -Recurse -Depth 3 | Where-Object { $_.PSIsContainer -eq $false -and $_.DirectoryName -notmatch ($excludeFolders -join "|") }
+  }
 
-    if ($projectData.project.psobject.properties.name -notcontains "artifactId" -or !$projectData.project.artifactId) {
-      Write-Host "$projectPath doesn't have a defined artifactId so skipping this pom."
-      return $null
+  foreach ($ymlFile in $ymlFiles)
+  {
+    # For each yml file do the following:
+    # 1. Load the yml file.
+    # 2. Get the Artifacts and AdditionalModules for the yml file.
+    # 3. Get the path to the yml
+    # 4. From the path in #1, recursively get a list of all pom.xml with depth 1. This
+    #    will take care of the cases where the pom.xml is sitting next to the ci.yml, in
+    #    the case where the ci.yml is in the sdk/<ServiceDirectory>/<LibraryDirectory>,
+    #    as well as the case where the ci.yml is in the sdk/<ServiceDirectory>
+    # 5. For each pom file, check and see if its group:artifact matches a name/groupId in
+    #    the ci.yml's Artifacts list.
+    # 6a. If #5 has a match, create the PackageProp and add it to the list
+    # 6b. If #5 doesn't have a match, then skip it. This is the case where it's either
+    #     an AdditionalModule or something from another track.
+    Write-Host "Processing $ymlFile"
```

Comment:
I worry about this function having too much noise in the logs in the various places this will be used. We should keep an eye on that and consider removing some of the logging or making it verbose logging.

Reply:
I don't think the verbosity is going to be a problem. If it becomes one, then I'll add something to make pieces less verbose.